README.md
# Temperature Express
![Temperature Express Logo](https://github.com/netbymatt/temperature-express/blob/main/html/images/tempexp-192.png?raw=true)

This is the base code for https://temperature.express

A graphical representation of the hourly forecast data and observations available through [https://api.weather.gov](https://www.weather.gov/documentation/services-web-api).

## Features
- Graph display of forcast for the next 7 days and observations for the previous ~7 days in an hour-by-hour format.
	- Temperature
	- Wind chill and heat index, shown as "Feels Like" as appropriate
	- Clouds
	- Precipitation: Rain, Snow, Ice
	- Wind
	- Dewpoint
	- Lightning
	- Barometer (observations only)
- GPS enabled to find your location and retreive a forcast
- Displays active alerts such as: Tornado, Severe Thunderstorm, Flash Flood
- Displays SPC Outlook when the location is inside of one of the highlighted risk areas
- Installable as a progressive web app (production site only)
- Alternative table view
- Sunrise and sunset times in graph and separately

## Source Data
The API for the National Weather Service has a huge amount of forecast and climate data for free including an hour-by-hour forecast for the next 7 days. Unfortunately, getting it displayed in a	concise	format is not something they do well. The main graph that this application produces attempts to display the most relevant parts of the forecast (high and low temperature, wind chill, heat index, clouds, precipitation and dewpoint) in an easy to digest format.

## Project Technical Info
In the production environment it is built to be very fast and to be installed as a [progressive web app](https://web.dev/progressive-web-apps/). The speed optimizations (service worker caching, minified javascript, etc) are not included in this repo as I would like this to also serve as a reasonably-sized web app for learning purposes.

# Usage
The project is a static set of HTML, Javascript and CSS files located in the ```html/``` folder. You can host them however you'd like. Keep in mind that there are CORS considerations when hosting this and that the NWS API does not always have the appropriate headers. It will not run by simply double-clicking the index.html file.

For simplicity, I've included a Node.js script that launches express web server and also passes through any requests to the NWS API so that everything can come from a single source, side-stepping the CORS issue.

``` bash
git clone git@github.com:netbymatt/temperature-express.git
npm i
node index.js
```
Open your web browser and point it at http<span>://</span>localhost:3000

If you're using VSCode, a launch.json file includes configurations for launching both the server and an instance of Chrome for client-side debugging.

# Donation
If you find this repo or https://temperature.express useful I would appreciate a little support to help pay for the cost of hosting the site.

<a href="https://www.buymeacoffee.com/temp.exp" target="_blank"><img src="https://www.buymeacoffee.com/assets/img/custom_images/orange_img.png" alt="Buy Me An Umbrella" style="height: 41px !important;width: 174px !important;box-shadow: 0px 3px 2px 0px rgba(190, 190, 190, 0.5) !important;-webkit-box-shadow: 0px 3px 2px 0px rgba(190, 190, 190, 0.5) !important;" ></a>

# Caution
The NWS API that supplies data for this project is still in active development and not all parts of it are considered production ready. It has a fairly reliable uptime, but this project should not be your only source of weather information during severe events.

# FAQ

Read the complete [FAQ](FAQ.md).

CHANGELOG.md
# Significant changes
Notable changes to the code and feature updates are listed below. View the commits to see all changes including bug fixes.

# Development
Changes listed below document the development of the software before the initial release. Some changes specific to the production version of temperature.express have been omitted.

## 1.19.4
- update about

## 1.19.3
- update loading phrases

## 1.19.2
- forecast, location refactor
- location refactor
- more refactoring
- alerts refactoring
- alerts refactor

## 1.19.1
- always scroll to bottom of console on load

## 1.19.0
- console copy and paste
- message formatting
- refactor progressbar
- display messages in gui
- better retry strategy throughout
- continue retrying load of forecast in background
- modularization

## 1.18.0
- modularize
- add webpack

## 1.17.8
- fix char visibility when loading
- modules complete
- initial conversion and test changes
- change global naming structure

## 1.17.7
- fix time zone offset for text forecast
- review foreach usage
- use forEachElem
- legend cleanup
- update luxon
- interface cleanup

## 1.17.6
- fix outlook/alert popups on top of each other

## 1.17.5
- fix loading throbber error

## 1.17.4
- separate chart marking
- relocate progress bar code

## 1.17.3
- update tooltip references

## 1.17.2
- refactor data organization
- refactor legend and loading
- refactor tooltip

## 1.17.1
- always hide text forecast when no message

## 1.17.0
- allow multiple lines
- styling
- make weather text strings
- correct 1px changeover error
- toggle area for text
- less code in formatting loop
- add lightning and null some zero values

## 1.16.2
- change barometer scaling

## 1.16.1
- fix wind gust fill color

## 1.16.0
- dark mode
- single scss file

## 1.15.0
- change to scss

## 1.14.2
- update spc ux

## 1.2.6

## 1.14.1
- update font awesome

## 1.14.0
- fix map dialog maximum width
- always list 3 spc outlook days

## 1.13.1
- close more dialogs automatically

## 1.13.0
- navigate through spc maps
- add spc maps

## 1.12.1
- Has warning only if active

## 1.12.0
- add wind gust
- wind direction in observations

## 1.11.0
- add barometer to forecast
- add barometer

## 1.10.1
- menu code cleanup
- location code cleanup
- holiday code cleanup
- astronomical code cleanup
- cleanup alert code
- forecast code cleanup
- modularize code

## 1.10.0
- remove location history from dialog

## 1.9.6
- red/yellow spc outlook button
- red/yellow alert button coloring

## 1.9.5
- update vendor dependencies

## 1.9.4
- dialog functionality cleanup

## 1.9.3
- properly handle holes in geojson
- update dependencies

## 1.9.2
- fix overflow css

## 1.9.1
- remove unnecessary console.log

## 1.9.0
- list history locations in menu

## 1.8.1
- nice show/hide of chart icons
- show icon animation

## 1.8.0
- add spc outlook

## 1.7.1
- make sunrise/sunset follow forecast location time zone

## 1.7.0
- graph in forecast location local time

## 1.6.7
- handle cancel updates to alert properly after supersceded was implemented

## 1.6.6
- add old data indication
- update dependencies

## 1.6.5
- fix days z-index

## 1.6.4
- add retry button

## 1.6.3
- allow days height to collapse

## 1.6.2
- switch to local urls

## 1.6.1
- annunciate error from nws

## 1.6.0
- track updated alerts
- add geolocation unavailable handling

## 1.5.4
- better progress bar, alerts failed indication

## 1.5.3
- correct handling when observations arrive first

## 1.5.2
- calculate oldest data even when observations are out of order
- update about

## 1.5.1
- fix vertical centering of loading icon

## 1.5.0
- remove jquery ui
- remove location from jquery dialog
- remove legend from jquery dialog
- drop dialog-error
- remove astronomical info from jquery dialog
- separate dialog css

## 1.4.1
- minify css

## 1.4.0
- add holidays

## 1.3.10
- hide table when changing location
- alert box hide/show animation
- remove jquery from astronomical info

## 1.3.9

## 1.3.8
- add loading transition

## 1.3.7
- remove jquery from loading
- remove jquery from legend
- remove jquery from days plotting
- hide tooltip on pan

## 1.3.6
- hide alert button on load
- remove more jquery
- more css transitions

## 1.3.5
- better tooltip
- update scaled-number dependencies

## 1.3.4
- code cleanup
- update dependencies

## 1.3.3
- side menu css fix
- update dependencies

## 1.3.2
- retry one time when nws api fails

## 1.3.1
- allow for no city name when using lat/lon lookup

## 1.3.0
- extract place name from open street map

## 1.2.11
- don't show null values in observations
- add gulp full_publish

## 1.2.10
- sort observation data

## 1.2.9
- fix css overflow

## 1.2.8
- fix alert box styling on mobile

## 1.2.7

## 1.2.6
- remove jquery from menu

## 1.2.5
- fix menu height issue

## 1.2.4
- sun/moon cleanup
- modernize alerts

## 1.2.3
- fix css compression

## 1.2.2
- linting cleanup
- cleanup scaled number
- upgrade luxon
- remove jquery from menu

## 1.2.1
- fix width and height on mobile

## 1.2.0
- remove jquery from alerts
- remove quirks mode
- add observation station
- switch to ejs
- update installability per pwa guidelines
- fix expires/end for alerts when one is not present
- change snow scale
- better error handling and code cleanup
- change ajax to async/await
- switch eslint styling
- change drawing of days to something more robust
- Source wind speed changed from m/s to km/h
- fix day of week
- Fix off by one day headers
- Update flot
- Add search history
- List history in location prompt
- History of places
- Check for fresh data on visibility event
- add location error messages
- clean start doesn't prompt user
- Hide address bar
- Added forecast to side menu
- Spelling
- Unit conversion complete
- Fix some table styling
- Add some style to the side menu
- Fix degree symbol
- Hide menu when opening alerts
- Add alerts to menu
- Change filtering ends/expires for alerts
- Remove focus from use gps
- Additional fixes
- Update menu options

html/resources/js/config.mjs
import { getOptions } from './options.mjs';
import ScaledNumber from '../vendor/scalednumber.mjs';

// returns an object representing how to plot each trend
const trendConfig = (displayName, scale, units, yAxis, lineType, valueFunction) => {
	const obj = {
		displayName,
		scale,
		units,
		yAxis,	// 0=temperature, 1=percent, 2=inches (precipitation)
		lineType, 	// 0=no fill, 1=fill to zero, 2=points only
		valueFunction,
	};
		// set default display units
	obj.scale.setUnit(units);
	return obj;
};

// returns line type based on plot type
const getLineType = (lineType, name, forceState = false) => {
	// get value from options, default to true if not yet present it options
	const visible = getOptions().visible[name];
	const opt = (visible !== undefined) ? visible ?? forceState : true;

	switch (lineType) {
	case 2:
		// points only
		return {
			show: false,
		};
	case 1:
		// filled under line to zero
		return {
			fill: opt,
			lineWidth: 0,
			show: opt,
		};
	case 0:
	default:
		// line only, no fill
		return {
			show: opt,
			lineWidth: 2,
		};
	}
};

// get point type
const getPointType = (pointType, name, forceState = false) => {
	// get value from options, default to true if not yet present it options
	const visible = getOptions().visible[name];
	const opt = (visible !== undefined) ? visible ?? forceState : true;

	switch (pointType) {
	case 2: {
		return {
			show: opt,
			fillColor: '#00000000',
		};
	}
	default:
		return {
			show: false,
		};
	}
};

const LEGEND_COLORS = {
	Temperature: 'rgb(203, 75, 75)',
	Dewpoint: 'rgb(123, 210, 98)',
	Precip: 'rgb(131, 194, 239)',
	Clouds: 'rgb(210, 210, 210)',
	'Feels Like': 'rgb(210, 98, 161)',
	'Wind Speed': 'rgb(237, 194, 64)',
	'Wind Gust': 'rgb(235, 150, 0)',
	Snow: 'rgb(133,133,239)',
	Ice: 'rgb(0,255,0)',
	Barometer: 'rgb(192,192,192)',
	'Lightning (1-5)': 'rgb(128,128,0)',
};

// return a set color for each legend
const colorByLegend = (legend, observation) => {
	// if observation change the opacity to 50% for faded look
	if (observation) {
		return LEGEND_COLORS[legend].replace(')', ', 0.5)');
	}
	return LEGEND_COLORS[legend];
};

const SCALES = {	// unit[0] = the units provided by the nws api (metric), [1] = US
	TEMPERATURE: [{ unit: '&deg;C' }, { unit: '&deg;F', m: 1 / 1.8, b: -32 / 1.8 }],
	WIND: [{ unit: 'km/h' }, { unit: 'mph', m: 1.61, b: 0 }],
	PERCENT: [{ unit: '%' }, { unit: '%', m: 1, b: 0 }],
	INCHES: [{ unit: 'mm' }, {
		unit: 'in', m: 25.4, b: 0, precision: 1,
	}],
	INCHES_ICE: [{ unit: 'mm' }, {
		unit: 'in', m: 25.4, b: 0, precision: 2,
	}],
	BAROMETER: [
		{ unit: 'Pa' },
		{
			unit: '"', m: 1 / 0.0002953, b: 0, precision: 2,
		}],
	LIGHTNING: [
		{ unit: '', m: 1, b: -1 },
		{ unit: '' },
		// {
		// 	unit: '', m: 1, b: 1, precision: 0,
		// },
	],
};

const AXIS = {
	PERCENT: 2,
	SNOW_IN: 3,
	ICE_IN: 4,
	BAROMETER_IN: 5,
	TEMPERATURE: 1,
	LIGHTNING: 6,
};

const FILL = {
	NO: 0,
	ZERO: 1,
	POINTS: 2,
};

// null zero values on forecast
const NULL_VALUES = (toNull = 0) => (value) => (value === toNull ? null : value);

const DAY_BG_COLORS = {
	FREEZING: {
		LIGHT: '#EEFFFF',
		DARK: '#0080FF30',
	},
	DAY: {
		LIGHT: '#F4F4F4F4',
		DARK: '#FFFFFF30',
	},
};

const AVAILABLE_TRENDS = { // trends that will be extracted from resulting data (forecast data)
	// display name, scaled number, display units, y axis, fill type, extra function
	skyCover: trendConfig('Clouds', new ScaledNumber(0, 0, 100, SCALES.PERCENT), '%', AXIS.PERCENT, FILL.ZERO),
	probabilityOfPrecipitation: trendConfig('Precip', new ScaledNumber(0, 0, 100, SCALES.PERCENT), '%', AXIS.PERCENT, FILL.ZERO, NULL_VALUES(0)),
	snowfallAmount: trendConfig('Snow', new ScaledNumber(0, 0, 1000, SCALES.INCHES), 'in', AXIS.SNOW_IN, FILL.ZERO, NULL_VALUES(0)),
	iceAccumulation: trendConfig('Ice', new ScaledNumber(0, 0, 1000, SCALES.INCHES_ICE), 'in', AXIS.ICE_IN, FILL.ZERO, NULL_VALUES(0)),
	temperature: trendConfig('Temperature', new ScaledNumber(0, -1000, 1000, SCALES.TEMPERATURE), '&deg;F', AXIS.TEMPERATURE, FILL.NO),
	dewpoint: trendConfig('Dewpoint', new ScaledNumber(0, -1000, 1000, SCALES.TEMPERATURE), '&deg;F', AXIS.TEMPERATURE, FILL.NO),
	apparentTemperature: trendConfig('Feels Like', new ScaledNumber(0, -1000, 1000, SCALES.TEMPERATURE), '&deg;F', AXIS.TEMPERATURE, FILL.NO),
	windSpeed: trendConfig('Wind Speed', new ScaledNumber(0, -1000, 1000, SCALES.WIND), 'mph', AXIS.PERCENT, FILL.NO),
	pressure: trendConfig('Barometer', new ScaledNumber(0, 0, 1e7, SCALES.BAROMETER), '"', AXIS.BAROMETER_IN, FILL.NO),
	lightningActivityLevel: trendConfig('Lightning (1-5)', new ScaledNumber(0, 0, 5, SCALES.LIGHTNING), '', AXIS.LIGHTNING, FILL.ZERO, NULL_VALUES(1)),
};

const AVAILABLE_OBS = { // observations that will be extracted from resulting data (past observations)
	// display name, scaled number, display units, y axis, fill type
	temperature: trendConfig('Temperature', new ScaledNumber(0, -1000, 1000, SCALES.TEMPERATURE), '&deg;F', AXIS.TEMPERATURE, FILL.NO),
	dewpoint: trendConfig('Dewpoint', new ScaledNumber(0, -1000, 1000, SCALES.TEMPERATURE), '&deg;F', AXIS.TEMPERATURE, FILL.NO),
	windSpeed: trendConfig('Wind Speed', new ScaledNumber(0, -1000, 1000, SCALES.WIND), 'mph', AXIS.PERCENT, FILL.NO),
	windGust: trendConfig('Wind Gust', new ScaledNumber(0, -1000, 1000, SCALES.WIND), 'mph', AXIS.PERCENT, FILL.POINTS),
	barometricPressure: trendConfig('Barometer', new ScaledNumber(0, 0, 1e7, SCALES.BAROMETER), '"', AXIS.BAROMETER_IN, FILL.NO),
	// precipitationLastHour: trendConfig('Precip', new ScaledNumber(0,-1000,1000,scales.PRECIP), 'mph', AXIS.SNOW_IN, FILL.NO),
	// special case in formatting routine to extract heat index or wind chill
	apparentTemperature: trendConfig('Feels Like', new ScaledNumber(0, -1000, 1000, SCALES.TEMPERATURE), '&deg;F', AXIS.TEMPERATURE, FILL.NO),
};

const AXIS_LIMITS = {
	y3: {
		// snow in inches
		min: 0,
		max: 8,
	},
	y4: {
		// ice in inches
		min: 0,
		max: 0.5,
	},
	y5: {
		// barmoeter in inches
		// max is intentionally high to keep the barometer towards the bottom of the graph
		min: 29,
		max: 34,
	},
};

export {
	getLineType,
	getPointType,
	colorByLegend,
	SCALES,
	AVAILABLE_TRENDS,
	AVAILABLE_OBS,
	DAY_BG_COLORS,
	AXIS_LIMITS,
};
